### ISSUE TITLE
Add the ability to pickle TPOTRegressor object

### ISSUE THREAD

**Issue by RafeyIqbalRahman** [EpistasisLab/tpot#1126]
The following text is more of a feature request than a bug report.

I have been using Flask to create a web app before trying TPOT. Flask requires a pickle file of the model to predict and display the result on the web app. However, since the TPOT object is not pickle-able, the web app isn't functional anymore.

It would be great if the ability to pickle the TPOTRegressor object is added.

**Comment by hanshupe** [comment]
I guess you can just pickle the final pipeline object (tpot.fitted_pipeline_), which is a normal sklearn pipeline and can be used independently from tpot.

**Comment by RafeyIqbalRahman** [comment]
> I guess you can just pickle the final pipeline object (tpot.fitted_pipeline_), which is a normal sklearn pipeline and can be used independently from tpot.

I tried but doing that returned an AttributeError.

**Comment by hanshupe** [comment]
Where did you get the error? I tried it and it worked.

**Comment by RafeyIqbalRahman** [comment]
> Where did you get the error? I tried it and it worked.

Can you show the code that works?

**Comment by hanshupe** [comment]
After fitting your tpot object you just call

pipeline_dump = pickle.dumps(tpot.fitted_pipeline_)
pipeline = pickle.loads(pipeline_dump)
print(pipeline)

**Comment by RafeyIqbalRahman** [comment]
Thanks for the code. It worked. But when I loaded the model on Flask, I'm getting a strange error. 

`ValueError: matmul: Input operand 1 has a mismatch in its core dimension 0, with gufunc signature`

The web app I am creating requires 2 parameters longitude and latitude to show prediction.

Here is the Colab notebook link: https://colab.research.google.com/drive/178OrUuWEigZC-y2A2O83AVDGZdwxIsoW?usp=sharing

**Comment by hanshupe** [comment]
How do you call the pipeline when you do predictions and what is the shape and type of your input data.

you can do predictions by executing pipeline.predict(data) where data can be a data frame or  matrix with the same number of columns as you used during training.

(Your link can't be accessed)

**Comment by RafeyIqbalRahman** [comment]
I called the pipeline using pickle.load. The shape of the data is 18x3. The datatype is float. The link is accessible now.
I'll try pipeline.predict to make predictions.

**Comment by hanshupe** [comment]
you called .predict(y) but you have to call .predict(X).

**Comment by RafeyIqbalRahman** [comment]
You are right. Actually, I was checking what is the reason for the error. Turns out that when I did .predict(y), I got the same error as I mentioned above. However, in the Flask code, I'm feeding X and still getting the same error as I got in .predict(y).

**Comment by hanshupe** [comment]
Like said above, the shape of your input data must match exactly the shape during training. You pass a list, so it's just 1 column with two rows. You need to do something like np.column_stack((a, b)), but there are different ways.

**Comment by RafeyIqbalRahman** [comment]
Can you show a practical example of reshaping the data? Or can you please do the same in the notebook's link so I can get an idea?

**Comment by hanshupe** [comment]
np.column_stack([list-variable])

**Comment by RafeyIqbalRahman** [comment]
This doesn't work. Also, I'm getting an error message saying that the features should be 2D while the target should be 1D.

**Comment by hanshupe** [comment]
just print the type and shape of your input. There are many threads on stackoverflow how you can convert it into 2D.

**Comment by RafeyIqbalRahman** [comment]
When converted into 2D, this is what I got.

![image](https://user-images.githubusercontent.com/59226057/95688672-4112b380-0c25-11eb-850f-26510126a4ca.png)

**Comment by RafeyIqbalRahman** [comment]
Also, as seen [here](https://stackoverflow.com/questions/55309176/how-to-correct-numpy-and-tpot-array-shapes-error), TPOT lacks multi-label regression ability.

**Comment by hanshupe** [comment]
Do you have multiple variables at output? If not then it must work. Just check the shapes of your input X and your target y.

**Comment by RafeyIqbalRahman** [comment]
No, I have a single variable at the output. The shape of X is (18, 2) and that of y is (18, 1).

**Comment by hanshupe** [comment]
But then it's not multi label.

**Comment by hanshupe** [comment]
Btw. I recommend that you secure your notebook again to not get security issues.

**Comment by RafeyIqbalRahman** [comment]
> But then it's not multi label.

So what's the solution now?

**Comment by hanshupe** [comment]
As explained, pickling is possible with tpot, also you don't have a multi label problem here, you just have to shape your data correctly.

**Comment by weixuanfu** [comment]
@hanshupe thank you for answer the question herein. 
@RafeyIqbalRahman  I think the `y`'s shape  should `(18, )` instead of `(18, 1)` if `y` is a 1-D array.

**Comment by RafeyIqbalRahman** [comment]
> @hanshupe thank you for answer the question herein.
> @RafeyIqbalRahman I think the `y`'s shape should `(18, )` instead of `(18, 1)` if `y` is a 1-D array.

I reshaped y using .flatten but still the Flask app is not showing the predicted result.

**Comment by weixuanfu** [comment]
@RafeyIqbalRahman I cannot check your Colab notebook (maybe permission issue?) and am not sure why Flask app is not working. Could you please provide a demo for reproducing the issue of pickling `tpot.fitted_pipeline_`?

**Comment by RafeyIqbalRahman** [comment]
@weixuanfu this is the link to my Colab notebook: https://colab.research.google.com/drive/1jVRhIZEV8rjdsQFPvWJof9R7wJcdF-cd?usp=sharing.  I feel there's some issue with the pickle file that's why the Flask app is not working.

**Comment by weixuanfu** [comment]
I have a quick look. I think `final_features` only has 1 feature but the model was fitted with 2 features (`X.shape=(18,2)`), which make the prediction did not work. You can add a line like `print(final_features.shape)` before `prediction = model.predict(final_features)` to check that.

**Comment by RafeyIqbalRahman** [comment]
> I have a quick look. I think `final_features` only has 1 feature but the model was fitted with 2 features (`X.shape=(18,2)`), which make the prediction did not work. You can add a line like `print(final_features.shape)` before `prediction = model.predict(final_features)` to check that.

I tried this but this did not work maybe because it is a function. Can you provide a function for getting the shape of array?

**Comment by RafeyIqbalRahman** [comment]
Thanks. Since final_features is a list object, I used len(final_features) to get the length and the length turned out to be 1 and the model is fitted with 2 features. How to solve this?

**Comment by RafeyIqbalRahman** [comment]
When I tried to reshape final_features, I got a ValueError saying that an array of size 1 cannot be reshaped into shape (1,2).

**Comment by weixuanfu** [comment]
I think the point here is not about reshaping the `final_features` and it should be about making `predict` function in the Flask to get the input X for prediction instead of making one from list object within the function. The input X should has a shape like (N, 2), N is the number of samples in the dataset.

**Comment by RafeyIqbalRahman** [comment]
So something in this part of the code needs to be changed.

![image](https://user-images.githubusercontent.com/59226057/96900045-189f7a80-14ab-11eb-83e7-4ead763035d3.png)
